doc/contributing.rst

```diff
@@ -1,221 +1,760 @@
+already-open PR. See the :ref:`testing <run-tests>` section (below) for
+examples of how to run tests locally, and make sure you have the testing
+dataset installed.
+
+
+Make tests fast and thorough
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Whenever possible, use the testing dataset rather than one of the sample
+datasets when writing tests; it includes small versions of most MNE-Python
+objects (e.g., :class:`~mne.io.Raw` objects with short durations and few
+channels). You can also check which lines are missed by the tests, then modify
+existing tests (or write new ones) to target the missed lines. Here's an
+example that reports which lines within ``mne.viz`` are missed when running
+``test_evoked.py`` and ``test_topo.py``:
 
 .. code-block:: console
 
-   $ pip install sphinx sphinx-gallery sphinx_bootstrap_theme sphinx_fontawesome memory_profiler pillow
+    $ pytest --cov=mne.viz --cov-report=term-missing mne/viz/tests/test_evoked.py mne/viz/tests/test_topo.py
 
-and then within the ``mne/doc`` directory do:
+You can also use ``pytest --durations=20`` to ensure new or modified tests will
+not slow down the test suite too much.
+
+
+Code style
+^^^^^^^^^^
+
+Adhere to standard Python style guidelines
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+All contributions to MNE-Python are checked against style guidelines described
+in `PEP 8`_. We also check for common coding errors (such as variables that are
+defined but never used). We allow very few exceptions to these guidelines, and
+use tools such as pep8_, pyflakes_, and flake8_ to check code style
+automatically. From the :file:`mne-python` root directory, you can check for
+style violations by running:
 
 .. code-block:: console
 
-   $ make html_dev-noplot
+    $ make flake
+
+in the shell. Several text editors or IDEs also have Python style checking,
+which can highlight style errors while you code (and train you to make those
+errors less frequently). This functionality is built-in to the Spyder_ IDE, but
+most editors have plug-ins that provide similar functionality. Search for
+:samp:`python linter <name of your favorite editor>` to learn more.
+
+
+Use consistent variable naming
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Classes should be named using ``CamelCase``. Functions and instances/variables
+should use ``snake_case`` (``n_samples`` rather than ``nsamples``). Avoid
+single-character variable names, unless inside a :term:`comprehension <list
+comprehension>` or :ref:`generator <tut-generators>`.
+
+
+Follow NumPy style for docstrings
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In most cases imitating existing docstrings will be sufficient, but consult the
+`Numpy docstring style guidelines`_ for more complicated formatting such as
+embedding example code, citing references, or including rendered mathematics.
+Private function/method docstrings may be brief for simple functions/methods,
+but complete docstrings are appropriate when private functions/methods are
+relatively complex.
```

Comment:
To run some basic tests on documentation, you can use:

$ pytest mne/tests/test_docstring_parameters
$ make docstyle